File: daemon/src/sallyport_daemon/protocol.py
"""Wire format shared with the extension.

Every WS frame is JSON with this shape:

    {v, ts, nonce, type, id?, body, mac}

where mac = HMAC-SHA256(secret, canonical_json({v, ts, nonce, type, id?, body}))
encoded as base64.

Both sides:
  * reject if `ts` differs from local clock by more than MAX_CLOCK_SKEW_S,
  * reject if `nonce` is reused (within NONCE_CACHE_SIZE),
  * reject on mac mismatch (constant-time compare).
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import math
import secrets as _secrets
import time
from collections import deque
from dataclasses import dataclass
from typing import Any

PROTOCOL_VERSION = 1
MAX_CLOCK_SKEW_S = 30
NONCE_CACHE_SIZE = 4096
# ECMAScript's Number::toString switches from positional digits to exponent
# notation at 1e21. Below that bound an integral double goes on the wire
# (JSON.stringify) as bare digits, so the canonical form must be those same
# digits; at/above it both languages use the scientific repr form.
_INTEGRAL_DIGITS_BOUND = 1e21


def _format_number(x: float) -> str:
    """Render a finite double exactly as the canonical form requires.

    * integral and |x| < 1e21 → the ECMAScript integer rendering (shortest
      round-trip digits, zero-padded) — exactly what JS ``String(x)`` /
      ``JSON.stringify`` produce, so the extension's wire frames re-parse
      to the same canonical bytes. For |x| ≤ 2^53−1 these are the exact
      digits. Folds ``-0.0`` to ``"0"``.
    * everything else → CPython ``repr`` (the extension reimplements this
      exact format as ``pythonFloatRepr``).
    """
    if x == 0:
        return "0"
    if x.is_integer() and abs(x) < _INTEGRAL_DIGITS_BOUND:
        r = repr(x)
        if "e" in r:
            # e.g. '1.152921504606847e+18' → '1152921504606847000'
            mantissa, _, exp = r.partition("e")
            neg = mantissa.startswith("-")
            digits = mantissa.lstrip("-").replace(".", "")
            padded = digits + "0" * (int(exp) - (len(digits) - 1))
            return "-" + padded if neg else padded
        # Fixed form: an integral repr always ends in '.0'.
        return r[:-2]
    return repr(x)


def _encode_canonical(value: Any) -> str:
    """Recursive canonical encoder. See :func:`canonical_json` for the rules."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ProtocolError("non-finite number in canonical JSON")
        return _format_number(value)
    if isinstance(value, int):
        # An int is wire-legal iff the peer's JSON.parse (which yields an
        # IEEE-754 double) preserves it: either the double equals the int
        # exactly, or the int IS the shortest rendering of that double
        # (i.e. it came off the wire from JSON.stringify). Anything else
        # would silently change value across the bridge — reject instead.
        try:
            as_float = float(value)
        except OverflowError as exc:
            raise ProtocolError("integer too large for IEEE-754 double") from exc
        rendered = _format_number(as_float)
        if int(as_float) != value and str(value) != rendered:
            raise ProtocolError("integer not exactly representable as IEEE-754 double")
        return rendered
    if isinstance(value, str):
        # json.dumps escapes exactly like before (and like JSON.stringify).
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(_encode_canonical(item) for item in value) + "]"
    if isinstance(value, dict):
        for key in value:
            if not isinstance(key, str):
                raise ProtocolError("non-string object key in canonical JSON")
        return (
            "{"
            + ",".join(
                json.dumps(key, ensure_ascii=False) + ":" + _encode_canonical(value[key])
                for key in sorted(value)
            )
            + "}"
        )
    raise ProtocolError(f"unsupported type in canonical JSON: {type(value).__name__}")


def canonical_json(value: Any) -> str:
    """Stable JSON encoder. Must match the extension's canonicalJson() byte-for-byte.

    Rules: keys sorted by Unicode code point, no whitespace, unicode
    passthrough. Numbers: integral doubles with |x| < 1e21 render as bare
    shortest digits (matching the extension's JSON.stringify wire output);
    all other finite doubles use CPython ``repr``; ints must be exactly
    double-representable (or be the shortest rendering of a double).
    Raises :class:`ProtocolError` on values that cannot round-trip
    identically through both languages (non-finite numbers, precision-losing
    ints, lone surrogates, non-string keys).
    """
    encoded = _encode_canonical(value)
    try:
        encoded.encode("utf-8")
    except UnicodeEncodeError as exc:
        # A lone surrogate cannot be UTF-8 encoded, so the peer could never
        # verify the MAC anyway. (JS strings can carry lone surrogates too;
        # canonicalJson rejects them symmetrically.)
        raise ProtocolError("lone surrogate in canonical JSON string") from exc
    return encoded


@dataclass
class Envelope:
    type: str
    body: Any
    id: str | None = None
    ts: int | None = None
    nonce: str | None = None
    v: int = PROTOCOL_VERSION


class ProtocolError(Exception):
    pass


class Signer:
    def __init__(self, secret: bytes) -> None:
        if len(secret) < 16:
            raise ValueError("secret too short (min 16 bytes)")
        self._secret = secret
        self._seen: deque[str] = deque(maxlen=NONCE_CACHE_SIZE)
        self._seen_set: set[str] = set()

    def _mac_input(self, env: dict[str, Any]) -> bytes:
        canonical = {
            "v": env["v"],
            "ts": env["ts"],
            "nonce": env["nonce"],
            "type": env["type"],
            "body": env["body"],
        }
        if env.get("id") is not None:
            canonical["id"] = env["id"]
        return canonical_json(canonical).encode("utf-8")

    def sign(self, env: Envelope) -> dict[str, Any]:
        ts = env.ts if env.ts is not None else int(time.time())
        nonce = (
            env.nonce
            if env.nonce is not None
            else base64.b64encode(_secrets.token_bytes(16)).decode("ascii")
        )
        payload: dict[str, Any] = {
            "v": env.v,
            "ts": ts,
            "nonce": nonce,
            "type": env.type,
            "body": env.body,
        }
        if env.id is not None:
            payload["id"] = env.id
        mac = hmac.new(self._secret, self._mac_input(payload), hashlib.sha256).digest()
        payload["mac"] = base64.b64encode(mac).decode("ascii")
        return payload

    def verify(self, raw: dict[str, Any]) -> Envelope:
        if not isinstance(raw, dict):
            raise ProtocolError("not an object")
        if raw.get("v") != PROTOCOL_VERSION:
            raise ProtocolError(f"bad version: {raw.get('v')!r}")
        ts = raw.get("ts")
        nonce = raw.get("nonce")
        typ = raw.get("type")
        mac_b64 = raw.get("mac")
        if not isinstance(ts, int):
            raise ProtocolError("bad ts")
        if not isinstance(nonce, str) or not nonce:
            raise ProtocolError("bad nonce")
        if not isinstance(typ, str) or not typ:
            raise ProtocolError("bad type")
        if not isinstance(mac_b64, str) or not mac_b64:
            raise ProtocolError("bad mac")
        if "body" not in raw:
            # _mac_input reads env["body"]; a missing key must be a clean
            # ProtocolError (skipped frame), not a KeyError teardown.
            raise ProtocolError("missing body")
        if "id" in raw and not isinstance(raw["id"], str):
            # The read loop uses id as a dict key (_pending.pop); an
            # unhashable id (dict/list) from an authenticated-but-buggy
            # peer must be a skipped frame, not a TypeError teardown.
            raise ProtocolError("bad id")

        now = int(time.time())
        if abs(now - ts) > MAX_CLOCK_SKEW_S:
            raise ProtocolError(f"timestamp skew {now - ts}s")

        if nonce in self._seen_set:
            raise ProtocolError("nonce replay")

        expected = hmac.new(self._secret, self._mac_input(raw), hashlib.sha256).digest()
        try:
            provided = base64.b64decode(mac_b64, validate=True)
        except Exception as exc:
            raise ProtocolError(f"mac not base64: {exc}") from exc
        if not hmac.compare_digest(expected, provided):
            raise ProtocolError("mac mismatch")

        # Evict the oldest entry explicitly *before* appending so that
        # set and deque stay in sync. (Relying on deque.append's silent
        # auto-eviction would pop from the wrong end of the set.)
        if len(self._seen) == self._seen.maxlen:
            evicted = self._seen.popleft()
            self._seen_set.discard(evicted)
        self._seen.append(nonce)
        self._seen_set.add(nonce)

        return Envelope(
            type=typ,
            body=raw.get("body"),
            id=raw.get("id"),
            ts=ts,
            nonce=nonce,
            v=raw["v"],
        )

File: extension/src/protocol.ts
// Wire protocol shared with the daemon.
//
// Every envelope is JSON of the form:
//   { v, ts, nonce, type, id?, mac, body }
// where `mac = HMAC-SHA256(secret, canonical_json({v, ts, nonce, type, id, body}))`
// encoded as base64. Both sides verify mac, reject if ts is older than
// MAX_CLOCK_SKEW_S, and reject if nonce has been seen recently.

export const PROTOCOL_VERSION = 1;
export const MAX_CLOCK_SKEW_S = 30;
export const NONCE_CACHE_SIZE = 4096;

export type ToolCallBody = {
  name: string;
  args: Record<string, unknown>;
};

export type ToolResultBody =
  { ok: true; data: unknown } | { ok: false; error: string; code?: string };

export type HelloBody = {
  extensionVersion: string;
};

export type EnvelopeIn =
  | { type: 'hello_ack'; body: Record<string, never> }
  | { type: 'ping'; body: Record<string, never> }
  | { type: 'pong'; body: Record<string, never> }
  | { type: 'tool_call'; id: string; body: ToolCallBody };

export type EnvelopeOut =
  | { type: 'hello'; body: HelloBody }
  | { type: 'ping'; body: Record<string, never> }
  | { type: 'pong'; body: Record<string, never> }
  | { type: 'tool_result'; id: string; body: ToolResultBody };

export type SignedEnvelope = {
  v: number;
  ts: number;
  nonce: string;
  type: string;
  id?: string;
  body: unknown;
  mac: string;
};

// ECMAScript's Number::toString switches from positional digits to exponent
// notation at 1e21. Below that bound an integral double goes on the wire
// (JSON.stringify) as bare digits, so the canonical form must be those same
// digits; at/above it both languages use the scientific repr form. The
// daemon pins the same bound (_INTEGRAL_DIGITS_BOUND in protocol.py).
const INTEGRAL_DIGITS_BOUND = 1e21;

// A high or low surrogate code unit outside a valid pair. Such a string has
// no UTF-8 encoding, so the daemon could never verify the MAC — reject at
// signing time instead of stranding the call (protocol.py rejects likewise).
const LONE_SURROGATE = /[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/;

// Sort by Unicode code point — what Python's sort_keys does. A plain
// Array.sort() compares UTF-16 code units, which puts astral-plane keys
// (surrogate pairs, 0xD800+) BEFORE U+E000..U+FFFF and diverges from Python.
function compareCodePoints(a: string, b: string): number {
  let i = 0;
  let j = 0;
  while (i < a.length && j < b.length) {
    const ca = a.codePointAt(i) as number;
    const cb = b.codePointAt(j) as number;
    if (ca !== cb) return ca - cb;
    i += ca > 0xffff ? 2 : 1;
    j += cb > 0xffff ? 2 : 1;
  }
  return a.length - i - (b.length - j);
}

// Format a float exactly like CPython's float.__repr__ (which json.dumps
// uses): shortest round-trip digits, fixed notation when the decimal
// exponent is in [-4, 16), otherwise scientific with an explicit sign and
// at least two exponent digits ('1e-07', '1.5e+22'). Only called for
// non-integral doubles (which always have a fractional digit, so exp10 <
// digits.length - 1) and integral doubles ≥ 1e21 (exp10 ≥ 21 → scientific):
// integral doubles below 1e21 take the bare-digits path in canonicalNumber.
function pythonFloatRepr(x: number): string {
  const neg = x < 0;
  // toExponential() without an argument yields the shortest digit string
  // that uniquely identifies the double — the same digits CPython derives.
  const [mantissa, expPart] = Math.abs(x).toExponential().split('e');
  const digits = mantissa.replace('.', '');
  const exp10 = parseInt(expPart, 10);
  let out: string;
  if (exp10 >= -4 && exp10 < 16) {
    if (exp10 >= 0) {
      out = digits.slice(0, exp10 + 1) + '.' + digits.slice(exp10 + 1);
    } else {
      out = '0.' + '0'.repeat(-exp10 - 1) + digits;
    }
  } else {
    const m = digits.length > 1 ? digits[0] + '.' + digits.slice(1) : digits;
    const sign = exp10 < 0 ? '-' : '+';
    out = m + 'e' + sign + String(Math.abs(exp10)).padStart(2, '0');
  }
  return neg ? '-' + out : out;
}

function canonicalNumber(x: number): string {
  if (!Number.isFinite(x)) {
    throw new Error('non-finite number in canonical JSON');
  }
  if (Number.isInteger(x) && Math.abs(x) < INTEGRAL_DIGITS_BOUND) {
    // Integral doubles below 1e21 canonicalise as bare digits — EXACTLY the
    // bytes JSON.stringify puts on the wire frame, so the daemon's re-parse
    // recanonicalises to the same MAC input. The daemon renders the same
    // shortest-digit form from CPython repr (protocol.py:_format_number).
    // Also folds -0 to '0', matching Python.
    return String(x);
  }
  return pythonFloatRepr(x);
}

function canonicalString(s: string): string {
  if (LONE_SURROGATE.test(s)) {
    throw new Error('lone surrogate in canonical JSON string');
  }
  return JSON.stringify(s);
}

// What JSON.stringify drops from objects (and nulls inside arrays). The wire
// frame is JSON.stringify(env), so the canonical bytes must agree with what
// the peer re-parses — otherwise the MAC never verifies.
function isOmitted(v: unknown): boolean {
  return v === undefined || typeof v === 'function' || typeof v === 'symbol';
}

// Canonical JSON: deterministic key ordering + number/string normalisation so
// both sides produce identical bytes. Mirrors protocol.py:canonical_json —
// change them together and regenerate fixtures/canonical-vectors.json.
// Throws on values that cannot round-trip identically through both languages
// (NaN/Infinity, lone surrogates).
export function canonicalJson(value: unknown): string {
  let v = value;
  // JSON.stringify honours toJSON (e.g. Date → ISO string); the wire frame
  // goes through JSON.stringify, so the MAC input must match.
  if (v !== null && typeof v === 'object') {
    const toJSON = (v as { toJSON?: unknown }).toJSON;
    if (typeof toJSON === 'function') v = (toJSON as () => unknown).call(v);
  }
  if (v === null) return 'null';
  switch (typeof v) {
    case 'boolean':
      return v ? 'true' : 'false';
    case 'number':
      return canonicalNumber(v);
    case 'string':
      return canonicalString(v);
    case 'object':
      break;
    default:
      throw new Error(`unsupported type in canonical JSON: ${typeof v}`);
  }
  if (Array.isArray(v)) {
    return '[' + v.map((el) => (isOmitted(el) ? 'null' : canonicalJson(el))).join(',') + ']';
  }
  const obj = v as Record<string, unknown>;
  const keys = Object.keys(obj)
    .filter((k) => !isOmitted(obj[k]))
    .sort(compareCodePoints);
  return '{' + keys.map((k) => canonicalString(k) + ':' + canonicalJson(obj[k])).join(',') + '}';
}
